keps/sig-api-machinery/2558-publish-version-openapi/README.md

```diff
@@ -0,0 +1,406 @@
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [x] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [x] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [x] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes a way for publishing feature gate and corresponding
+versioning information in OpenAPI using comment tags on API types.
+
+Specifically, the versioning information would include the
+prerelease status, minimum Kubernetes version for the prerelease status
+and the feature gate associated with the struct or field.
+
+## Motivation
+
+**Feature gate**
+
+When a new field is added to an existing API version, a feature gate
+is added to control the enablement of the new field. Sometimes feature
+gates are also used when a new `Kind` is added.
+
+The feature gates are defined in [`kube_features.go`]. The comments
+for each of these feature gates in [`kube_features.go`] lists
+the prerelease status of the feature gate and the minimum Kubernetes
+version for the particular prerelease. These comments do not follow
+a pattern and are not parseable.
+
+Currently, when a new field is added, there is no way to map and document
+the field with the feature gate that controls it. Some fields mention
+the feature gates in a comment but these comments are not parseable either.
+
+**Field version**
+
+Currently, when a new field is added to a stable API version, there
+is no way to expose the version of the new field in OpenAPI.
+One way to identify alpha API fields is by comparing the OpenAPI schema
+generated from an apiserver with all APIs enabled and the one generated
+from an apiserver with only stable and beta APIs enabled.
+However, this process is cumbersome and not always feasible because not all
+fields are gated at the API group level.
+
+Since the OpenAPI definition is the source of truth for API documentation,
+this KEP proposes a way to publish versioning and feature gate
+information via OpenAPI.
+
+As a follow-up, this information can also be used to:
+- update the Kubernetes API reference documentation
+- surface [server-side warnings] in client-go and kubectl when a deprecated
+field or alpha field is being used.
+
+### Goals
+
+- Publish feature gate associated with an API type for built-in API types
```

Comment:
A lot of places in this doc mention "API type", but IIUC the tag is at field-level, and you don't plan to support type-level tag. Is that correct?